src/rdkafka_cgrp.c

```diff
@@ -3169,9 +4158,270 @@ rd_kafka_cgrp_max_poll_interval_check_tmr_cb (rd_kafka_timers_t *rkts,
+                             "Group \"%.*s\": postponing join until "
+                             "up-to-date metadata is available",
+                             RD_KAFKAP_STR_PR(rkcg->rkcg_group_id));
+
+                /* Join will occur after metadata refresh completes */
+                return RD_KAFKA_RESP_ERR_NO_ERROR;
+        }
+
+        rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER, "SUBSCRIBE",
+                     "Group \"%.*s\": modifying subscription of size %d to "
+                     "new subscription of size %d%s (join state %s)",
+                     RD_KAFKAP_STR_PR(rkcg->rkcg_group_id),
+                     rkcg->rkcg_subscription->cnt, rktparlist->cnt,
+                     !unsubscribing_topics ? " without removing any topics" :
+                     (!revoking ? " without revoking any partitions" : ""),
+                     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state]);
+
+        /* Create a list of the topics in metadata that matches the new
+         * subscription */
+        tinfos = rd_list_new(rkcg->rkcg_subscription->cnt,
+                             (void *)rd_kafka_topic_info_destroy);
+
+        /* Unmatched topics will be added to the errored list. */
+        errored = rd_kafka_topic_partition_list_new(0);
+
+        if (rkcg->rkcg_flags & RD_KAFKA_CGRP_F_WILDCARD_SUBSCRIPTION)
+                rd_kafka_metadata_topic_match(rkcg->rkcg_rk,
+                                              tinfos, rkcg->rkcg_subscription,
+                                              errored);
+        else
+                rd_kafka_metadata_topic_filter(rkcg->rkcg_rk,
+                                               tinfos,
+                                               rkcg->rkcg_subscription,
+                                               errored);
+
+        /* Propagate consumer errors for any non-existent or errored topics.
+         * The function takes ownership of errored. */
+        rd_kafka_propagate_consumer_topic_errors(
+                rkcg, errored, "Subscribed topic not available");
+
+        if (rd_kafka_cgrp_update_subscribed_topics(rkcg, tinfos)) {
+
+                if (!revoking) {
+                        rd_kafka_cgrp_rejoin(rkcg);
+                        return RD_KAFKA_RESP_ERR_NO_ERROR;
+                }
+
+                rd_kafka_dbg(rkcg->rkcg_rk,
+                        CONSUMER|RD_KAFKA_DBG_CGRP,
+                        "REBALANCE", "Group \"%.*s\" revoking "
+                        "%d of %d partition(s)",
+                        RD_KAFKAP_STR_PR(rkcg->rkcg_group_id),
+                        revoking->cnt,
+                        rkcg->rkcg_assignment->cnt);
+
+                rd_kafka_rebalance_op_incr(rkcg,
+                        RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS,
+                        revoking, rd_true/*rejoin*/, "subscribe");
+
+                rd_kafka_topic_partition_list_destroy(revoking);
+        }
+
```

Comment:
is revoking leaked here?

Reply:
I don't think so, because I think `rd_kafka_cgrp_update_subscribed_topics` would always return true if revoking not empty set. However, your point still holds - this takes too much effort to read. I have re-written the logic to be more robust (took the revoking case out of the above if).